pylint/checkers/unicode.py

```diff
@@ -0,0 +1,670 @@
+# Copyright (c) 2021 Carli Freudenberg <carli.freudenberg@energymeteo.de>
+
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+"""Unicode and some other ASCII characters can be used to create programs that run
+much different as a human reader would expect it.
+
+PEP 672 lists some examples.
+
+The following checkers are intended to make users aware of these issues.
+
+Some of the test might be too strict for users programming in other languages as
+english. As we assume this is minority than can disable the given tests, this should
+be not an issue.
+
+Note that the codes try to be prepared for UTF-16 and UTF-32 even so Python files
+can't be encoded in them yet.
+There is a checker that would detect UTF-16/UTF-32 files and fail, so even the usage
+of an old pylint version would not cause a "security hole" when used with new Python
+version that allow these encodings.
+"""
+import codecs
+import re
+import sys
+from collections import OrderedDict
+from functools import lru_cache
+from tokenize import detect_encoding
+from typing import Dict, List, NamedTuple, Optional, Sequence, TypeVar, Union
+
+from astroid import nodes
+
+import pylint.checkers.utils as pylint_utils
+from pylint import interfaces
+from pylint.checkers import BaseChecker
+from pylint.lint import PyLinter
+
+if sys.version_info[:2] >= (3, 8):
+    from typing import Protocol
+else:
+
+    class Protocol:
+        ...
+
+
+if sys.version_info[:2] >= (3, 7):
+    # Once we require at least Python 3.7, we can remove this lines
+    Py37Str = str
+else:
+
+    class Py37Str(str):
+        # Allow Python 3.6 compatibility
+        def isascii(self: str) -> bool:
+            return all("\u0000" <= x <= "\u007F" for x in self)
+
+
+# pylint disable=W0511
+
+# Used by ASCII Only checker, based on pylint.checkers.base.HUMAN_READABLE_TYPES
+# but limited to reduce complexity
+HUMAN_READABLE_TYPES = {
```

Comment:
This could be moved to `pylint.constants`. No need to duplicate the code here and it will benefit form any future additions.